File: Python/String_to_Integer_atoi.py
# leetcode problem link: https://leetcode.com/problems/string-to-integer-atoi/

class Solution:
    def myAtoi(self, str: str) -> int:
        num, i = [], 0
        while i < len(str) and str[i] == ' ':
            i += 1
        if i < len(str) and (str[i] == '-' or str[i] == '+'):
            num.append(str[i])
            i += 1
        while i < len(str) and str[i].isdigit():
            num.append(str[i])
            i += 1
        if len(num) == 0:
            return 0
        if not num[-1].isdigit():
            return 0
        ans = int(''.join(num))
        if ans < -(1 << 31):
            return -(1 << 31)
        if ans > (1 << 31) - 1:
            return (1 << 31) - 1
        return ans

File: C++/String_to_integer_atoi.cpp
// https://leetcode.com/problems/string-to-integer-atoi/
class Solution {
   public:
    int myAtoi(string str) {
        int sign = 1, i = 0;
        while (str[i] == ' ') i++;
        if (str[i] == '-' || str[i] == '+') sign = 1 - 2 * (str[i] == '-'), i++;
        long long int base = 0;
        while (str[i] >= '0' && str[i] <= '9') {
            base = 10 * base + (str[i++] - '0');
            if (base > INT_MAX) {
                if (sign == -1) return INT_MIN;
                return INT_MAX;
            }
        }
        if (base > INT_MAX) {
            if (sign == -1)
                return INT_MIN;
            else
                return INT_MAX;
        }
        return (int)base * sign;
    }
};